masking/mask_terms.py

```diff
@@ -16,97 +15,186 @@ def is_comment_or_empty(s: str) -> str:
+                    
+                if term in self.terms:
+                    self.counts_dupes[term] += term
+                else:
+                    self.terms[term] = (translation, label)
+
+    def load_patterns(self, file: str, label_override: Optional[str] = None):
+        with open(file) as infh:
+            for line in infh:
+                if is_comment_or_empty(line):
+                    continue
+
+                elements = line.rstrip().split('|||')
+                if len(elements) < 1:
+                    raise Exception('Invalid pattern file: all lines must have a label')
+                pattern = elements[0].strip()
+                if label_override:
+                    label = label_override
+                else:
+                    label = elements[1].strip()
+                
+                # Boundary checking also needs to be handled in the patterns themselves
+                # because the behavior is different with word/non-word characters
+                # on the edges!
+                #pattern = ' ' + pattern + ' ' 
                 self.patterns.append((pattern, label))
 
-    def unmask(self, output, orig_source, masked_source):
+    def get_mask_string(self, label, index: Optional[int] = None):
+        """
+        Produces a maybe-indexed mask.
+        """
+        if index is None:
+            return ' __{}__ '.format(label)
+        else:
+            return ' __{},{}__ '.format(label, index)
+    
+    def unmask(self, output, masks):
         """
         Removes masks.
 
-        orig_source: The boy is 10
-        masked_source: The boy is __NUM,1__
-        output: Le garçon est __NUM,1__
-
-        mask2word: { '__NUM,1__': '10' }
-
         """
-        # Create a dictionary mapping indexed masks to their corresponding unmasked source words
-        mask2word = dict(filter(lambda x: self.mask_matcher.match(x[0]), zip(masked_source.split(), orig_source.split())))
-        # Replace these in the string one by one
-        output = ' '.join([mask2word.get(word, word) for word in output.split()])
-
-        return output
+        unmasked = output
+        for mask in masks:
+            maskstr = mask["maskstr"]
+            #matched = mask["matched"]
+            replacement = mask["replacement"]
+            unmasked = unmasked.replace(maskstr,replacement)
+        
+        return unmasked
 
     def mask(self, orig_source, orig_target: Optional[str] = None):
-        # pad with spaces
-        source = ' {} '.format(orig_source).replace(' ', '  ')
-        target = None if orig_target is None else ' {} '.format(orig_target)
-
-        # increment the indices to use
-        indices = defaultdict(int)
+        masked_source, masked_target, term_masks = self.mask_by_term(orig_source, orig_target)
+        masked_source, masked_target, pattern_masks = self.mask_by_pattern(masked_source, masked_target)
+        term_masks.extend(pattern_masks)
```

Comment:
maybe a comment here describing what you're doing? (adding indexes afterwards, right?)